crates/iota-indexer/src/indexer_reader.rs

```diff
@@ -225,6 +226,109 @@ impl IndexerReader {
         Ok(stored_object)
     }
 
+    /// Fetches a past object by its ID and version.
+    ///
+    /// - If `before_version` is `false`, it looks for the exact version.
+    /// - If `true`, it finds the latest version before the given one.
+    ///
+    /// Searches the requested object version and checkpoint sequence number
+    /// in `objects_version` and fetches the requested object from
+    /// `objects_history`.
+    pub(crate) async fn get_past_object_read(
+        &self,
+        object_id: ObjectID,
+        object_version: SequenceNumber,
+        before_version: bool,
+    ) -> Result<PastObjectRead, IndexerError> {
+        let object_id_bytes = object_id.to_vec();
+        let object_version_num = object_version.value() as i64;
+
+        // Query objects_version to find the requested version and relevant
+        // checkpoint sequence number considering the `before_version` flag.
+        let pool = self.get_pool();
+        let object_id_bytes_clone = object_id_bytes.clone();
+        let object_version_info: Option<StoredObjectVersion> =
+            run_query_async!(&pool, move |conn| {
+                let version_condition: Box<
+                    dyn BoxableExpression<objects_version::table, diesel::pg::Pg, SqlType = Bool>,
+                > = if before_version {
+                    Box::new(objects_version::object_version.lt(object_version_num))
+                } else {
+                    Box::new(objects_version::object_version.eq(object_version_num))
+                };
+
+                objects_version::dsl::objects_version
+                    .filter(objects_version::object_id.eq(object_id_bytes_clone))
+                    .filter(version_condition)
+                    .order_by(objects_version::object_version.desc())
+                    .limit(1)
+                    .first::<StoredObjectVersion>(conn)
+                    .optional()
+            })?;
+
+        let Some(object_version_info) = object_version_info else {
+            // Check if the object ever existed.
+            let pool = self.get_pool();
+            let object_id_bytes_clone = object_id_bytes.clone();
+            let latest_existing_version: Option<i64> = run_query_async!(&pool, move |conn| {
+                objects_version::dsl::objects_version
+                    .filter(objects_version::object_id.eq(object_id_bytes_clone))
```

Comment:
Nit: Avoid the `_clone` suffix
```suggestion
            let object_id_bytes = object_id.to_vec();
            let latest_existing_version: Option<i64> = run_query_async!(&pool, move |conn| {
                objects_version::dsl::objects_version
                    .filter(objects_version::object_id.eq(&object_id_bytes))
```